src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,458 @@
+#define BUFFER_FAIL FAIL
+
+static int32_t PopulateByteArray(JNIEnv *env, jbyteArray source, uint8_t *dest, int32_t len);
+static int32_t PopulateString(JNIEnv *env, jstring source, char *dest, int32_t len);
+
+// Handles both DER and PEM formats
+jobject /*X509Certificate*/ AndroidCryptoNative_DecodeX509(const uint8_t *buf, int32_t len)
+{
+    assert(buf != NULL && len > 0);
+    JNIEnv* env = GetJNIEnv();
+
+    jobject ret = NULL;
+    INIT_LOCALS(loc, bytes, stream, certType, certFactory)
+
+    // byte[] bytes = new byte[] { ... }
+    // InputStream stream = new ByteArrayInputStream(bytes);
+    loc[bytes] = (*env)->NewByteArray(env, len);
+    (*env)->SetByteArrayRegion(env, loc[bytes], 0, len, (const jbyte*)buf);
+    loc[stream] = (*env)->NewObject(env, g_ByteArrayInputStreamClass, g_ByteArrayInputStreamCtor, loc[bytes]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+    // return (X509Certificate)certFactory.generateCertificate(stream);
+    loc[certType] = JSTRING("X.509");
+    loc[certFactory] = (*env)->CallStaticObjectMethod(env, g_CertFactoryClass, g_CertFactoryGetInstance, loc[certType]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = (*env)->CallObjectMethod(env, loc[certFactory], g_CertFactoryGenerateCertificate, loc[stream]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    if (ret != NULL)
+        ret = ToGRef(env, ret);
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+    return ret;
+}
+
+// Encodes as DER format
+int32_t AndroidCryptoNative_EncodeX509(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+    int32_t ret = BUFFER_FAIL;
+
+    // byte[] encoded = cert.getEncoded();
+    // return encoded.length
+    jbyteArray encoded =  (*env)->CallObjectMethod(env, cert, g_X509CertGetEncoded);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = PopulateByteArray(env, encoded, buf, len);
+
+cleanup:
+    (*env)->DeleteLocalRef(env, encoded);
+    return ret;
+}
+
+void CryptoNative_X509Destroy(jobject /*X509Certificate*/ cert)
+{
+    ReleaseGRef(GetJNIEnv(), cert);
+}
+
+jobject /*X509Certificate*/ CryptoNative_X509UpRef(jobject /*X509Certificate*/ cert)
+{
+    return AddGRef(GetJNIEnv(), cert);
+}
+
+int64_t AndroidCryptoNative_X509GetNotAfter(jobject /*X509Certificate*/ cert)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+    int64_t ret = VALUE_FAIL;
+
+    // Date notAfter = cert.getNotAfter()
+    // return notAfter.getTime()
+    jobject notAfter = (*env)->CallObjectMethod(env, cert, g_X509CertGetNotAfter);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = (int64_t)(*env)->CallLongMethod(env, notAfter, g_DateGetTime);
+
+cleanup:
+    (*env)->DeleteLocalRef(env, notAfter);
+    return ret;
+}
+
+int64_t AndroidCryptoNative_X509GetNotBefore(jobject /*X509Certificate*/ cert)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
```

Comment:
How expensive is this?  If it's actually somewhat expensive would it make more sense to squeeze as many of these items into as few calls as possible?

e.g.

```C
int32_t AndroidCryptoNative_X509GetSimpleValues(
    jobject /*X509Certificate*/ cert,
    int64_t* pNotBeforeOut,
    int64_t* pNotAfterOut,
    jobject* /*X500Principal**/ pSubjectNameOut,
    etc);
```

If it's not expensive then the one call per item probably makes lifetime and errors easier to manage.